Adxrs290 Buffered channel support.

kister-jimenez opened the pull request:
To make this tinyiio demo as close as possible to the linux version, buffered channels has been added. The following are added:
1. Added buffered channel reading in the iio\iio_adxrs290 and the project.
2. Added the sync pin support to the driver to be able to check if the device has data available. 
3. Added a data_ready function from the driver and use this when getting data for the read_dev. This enables the read_dev to get the data only if it is available. Since TRIGGER is not yet supported by tinyiio, this acts as the throttling mechanism to ensure a minimum data reading interval.

This is how it will look like in IIO-OSC capture window (Auto scale disabled: Ymax: 30000, Ymin:-30000):
![image](https://user-images.githubusercontent.com/60493211/104317361-d3726e00-5518-11eb-89cc-ffaf44f003b2.png)

mchindri commented on iio/iio_adxrs290/iio_adxrs290.c:
```diff
@@ -178,3 +179,61 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
+		while (get_next_ch_idx(adxrs290_ch_mask, ch_idx, &ch_idx)) {
+			adxrs290_get_rate_data(dev, ch_idx, &data);
+			if (ch_idx == ADXRS290_CHANNEL_TEMP)
+				data = (data << 4) >> 4;
```
nitpick: data &= 0xFFF0

kister-jimenez replied:
This is for sign-extension of temp value. The << will cause it to lose the 4 MSB. In the >> it go back to original form but MSB is now sign extended.

mchindri replied:
sorry: I ment data &= 0xFFF

mchindri commented on drivers/gyro/adxrs290/adxrs290.c:
```diff
@@ -292,6 +292,13 @@ int32_t adxrs290_init(struct adxrs290_dev **device,
 		goto error_spi;
 
+	// Set GPIO sync pin. Set adxrs290 to output on sync pin.
+	gpio_get(&dev->gpio_sync, &init_param->gpio_sync);
```
gpio_remove is needed in adxr290_remove

mchindri replied:
Also check return values

mchindri commented on drivers/gyro/adxrs290/adxrs290.c:
```diff
@@ -292,6 +292,13 @@ int32_t adxrs290_init(struct adxrs290_dev **device,
+	// Set GPIO sync pin. Set adxrs290 to output on sync pin.
+	gpio_get(&dev->gpio_sync, &init_param->gpio_sync);
+	gpio_direction_input(dev->gpio_sync);
+	ret |= adxrs290_reg_write(dev, ADXRS290_REG_DATA_READY, ADXRS290_DATA_RDY_OUT);
```
is this related to this commit?

kister-jimenez replied:
Yes. The GPIO needs to be initialized as input and the device should be configured to output the DATA_READY on the sync pin. This DATA ready is needed so that it can be used in the filling of buffers in the iio_demo in the next commits.

mchindri commented on projects/adxrs290-pmdz/src/main.c:
```diff
@@ -156,6 +159,12 @@ int main(void)
 
+	/* Aducm platform dependent initialization for GPIO. */
+	gpio_sync_init_param = (struct gpio_init_param) {
+		.number = 0x10,
```
Use better a define for this value

kister-jimenez commented:
V2:
- define GPIO_SYNC_PIN_NUM 0x10 for the aducm3029 sync pin.
- Moved the gpio initialization in the main outside the ifdef ADUCM_PLATFORM since it's generic now.
- Checked returns of gpio routines and added gpio_remove() in the adxrs290_remove() and a goto label for error_gpio just before the error_spi.

amiclaus commented on iio/iio_adxrs290/iio_adxrs290.c:
```diff
@@ -178,3 +179,67 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
+	}
+
+	return nb_samples;
+}
```
nitpick: new line at end of file

kister-jimenez replied:
Done.

adrimbarean commented on iio/iio_adxrs290/iio_adxrs290.c:
```diff
@@ -178,3 +179,57 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
+	return 1;
+}
+
+int32_t	adxrs290_read_samples(void *device, uint16_t *buff,
```
nitpick: tab after int32_t

kister-jimenez replied:
Done.

kister-jimenez commented on iio/iio_adxrs290/iio_adxrs290.c:
```diff
@@ -49,6 +49,7 @@
 #define MAX_REG_ADDR		0x12
 
 volatile uint8_t current_direct_reg = 0;
+volatile uint8_t adxrs290_ch_mask = 0;
```
Removed the remnants of the register routines.
Removed global variables and moved it to driver as a field of adxrs290. Added a routine to read all data in one SPI transaction.

kister-jimenez commented on iio/iio_adxrs290/iio_adxrs290.h:
```diff
@@ -133,10 +152,9 @@ struct iio_device adxrs290_iio_descriptor = {
-	.read_data = NULL,
-	.write_data = NULL,
+	.prepare_transfer = adxrs290_update_active_channels,
+	.end_transfer = adxrs290_close_channels,
```
Done.

kister-jimenez commented:
V3:
- Removed the global variables and other unused variables and prototypes in the iio_demo of adxrs290.
- Added the routine that can get all data in one SPI transaction to ensure that no data has changed  between multiple byte reads. [Datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/ADXRS290.pdf) section RATE OUTPUT DATA.

mchindri commented on iio/iio_adxrs290/iio_adxrs290.c:
```diff
@@ -173,3 +173,30 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
+
+	offset = 0;
+	for (i = 0; i < nb_samples; i++) {
+		adxrs290_get_burst_data(dev, data, &ch_cnt);
```
```suggestion
		adxrs290_get_burst_data(dev, data + offset, &ch_cnt);
```
And you will not need memcpy

mchindri commented on drivers/gyro/adxrs290/adxrs290.c:
```diff
@@ -346,13 +369,30 @@ int32_t adxrs290_init(struct adxrs290_dev **device,
 		goto error_spi;
 
+	// Set GPIO sync pin.
+	ret |= gpio_get(&dev->gpio_sync, &init_param->gpio_sync);
```
if you check every time I think is better to use = instead of |=
